### PR TITLE
Custom association keys

### PR DESCRIPTION
**Pull request by 2468ben**
Here is the code I've changed so far to get a bunch acceptance tests working with #266 and #270.
Most of it is allowing the custom association keys, and the rest are little bits I'll comment on in the code.

I know you're moving and refactoring heavily right now, hopefully some of this is useful if you wanted to fix these things a better way.

### CODE REVIEW

File: addon/orm/associations/belongs-to.js [hunk 1]
```diff
@@ -72,15 +74,15 @@ class BelongsTo extends Association {
           - sets the associated parent (via model)
       */
       set: function(newModel) {
-        if (newModel && newModel.isNew()) {
-          this[foreignKey] = null;
+        if (newModel && newModel['isNew'] && newModel.isNew()) {
```
Comment (2468ben) [inline addon/orm/associations/belongs-to.js]: the `newModel` that's getting sent in (at least sometimes) is a stripped down POJO, so it couldn't call isNew().

File: addon/orm/associations/has-many.js [hunk 2]
```diff
@@ -114,11 +114,11 @@ class HasMany extends Association {
           schema[association.target].where(query).update(foreignKey, null);
 
           // Save any children that are new
-          models.filter(model => model.isNew())
+          models.filter(model => model['isNew'] && model.isNew())
             .forEach(model => model.save());
 
           // Associate the new children to this model
-          schema[association.target].find(models.map(m => m.id)).update(foreignKey, this.id);
+          schema[association.target].find(models.map(m => m.id || m.attrs.id)).update(foreignKey, this.id);
```
Comment (2468ben) [inline addon/orm/associations/has-many.js]: Some times that it was called I had to grab the attrs hash for the id. I'm really sorry but I didn't dig back to find the cause of it.

File: addon/orm/model.js [hunk 3]
```diff
@@ -178,7 +180,7 @@ class Model {
     Object.keys(this.belongsToAssociations).forEach(key => {
       var association = this.belongsToAssociations[key];
       var parent = this[key];
-      if (parent.isNew()) {
+      if (parent && parent.isNew()) {
```
Comment (2468ben) [inline addon/orm/model.js]: null check (for when the belongsTo relationship is empty/async).

File: addon/serializer-registry.js [hunk 4]
```diff
@@ -96,21 +96,22 @@ export default class SerializerRegistry {
     // Traverse this model's relationships
     let serializer = this._serializerFor(model);
     serializer.relationships
-      .map(key => model[key])
-      .forEach(relationship => {
+      .filter(key => model[key])
+      .forEach(relationshipKey => {
+        let relationship = model[relationshipKey];
```
Comment (2468ben) [inline addon/serializer-registry.js]: changed so I can access the relationshipKey and pass it through to the functions below.

**Comment by 2468ben** [comment]
I ran into another definedProperty taking over when the orm is re-initialized, and couldn't track down the cause so I had to move back off the orm layer for now. I'd be glad to slack chat on anything I found over the past ~week squeezing my API into it, and/or help out with wherever you plan to go.
I've spent way too much of my ember life on API serializing/deserializing before mirage, so my biggest feature request is a time machine so we could use it in 2013.

**Comment by samselikoff** [comment]
Thanks for looking into this, I'll circle back after I tie up the serializer stuff. I didn't expect anyone to be using the orm yet, I wanted to finish serializers/factories first and then document it. Thanks for giving it a try.

**Comment by 2468ben** [comment]
No problem, if you're still ironing out serializers there's only one 'feature' would love to sneak in:

Right now if I add a custom `Serialize.serialize(record, request)`, the request doesn't get passed in yet. If I had that AND a similar function to pick which relationships are included (like `Serialize.getRelationshipKeys(record, request)1, I could toss out all my homegrown serializers.

For example an endpoint like this:
`GET /comments?include_user=true`
```
//serializers/comment.js
getRelationshipKeys(record, request) {
   return request.queryParams.include_replies ? ['user','replies'] : ['replies'];
}
```

**Comment by samselikoff** [comment]
I definitely want to an api that supports inline serializers, which (I think) would let you accomplish that:

```js
this.get('/comments', function(schema, request) {
  let includeReplies = request.queryParams && request.queryParams.include_replies;
  let comments = schema.comment.all();

  return new Mirage.Response({
    content: comments,
    serializer: CommentSerializer.extend({
      relationships: includeReplies ? ['user', 'replies'] : ['replies]
    });
});
```

Another option will always be to just return JSON from your route, which will pass through the serializer layer:

```js
this.get('/comments', function(schema, request) {
  let includeReplies = request.queryParams && request.queryParams.include_replies;
  let comments = schema.comment.all();
  let SpecialSerializer = CommentSerializer.extend({
    relationships: includeReplies ? ['user', 'replies'] : ['replies]
  });

  return new SpecialSerializer().serialize(comments);
});
```

Finally we could also add the option for `relationships` to take a function, which gives you the request:

```js
// serializers/comment.js
export default Serializer.extend({
  relationships(request) {
    let includeReplies = request.queryParams && request.queryParams.include_replies;
    relationships: includeReplies ? ['user', 'replies'] : ['replies];
  }
});
```

**Comment by 2468ben** [comment]
Closing this because #266 and #270 are fixed and these changes aren't needed

**Pull request closed without merging**